core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -251,9 +251,16 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
               type.getFieldNames()),
           type.isNullable());
     } else if (type instanceof JavaType) {
-      return typeFactory.createTypeWithNullability(
-          typeFactory.createSqlType(type.getSqlTypeName()),
-          type.isNullable());
+      SqlTypeName sqlTypeName = type.getSqlTypeName();
+      final RelDataType relDataType;
+      if (sqlTypeName == SqlTypeName.ARRAY) {
+        RelDataType elementType = type.getComponentType() == null
+            ? typeFactory.createSqlType(SqlTypeName.ANY) : type.getComponentType();
```

Comment:
Does it need special consideration when it's a nested array, i.e. the element type of array is also of array

Reply:
Could you please give an example of table function for which it's necessary to accept nested array as argument ?

Reply:
You can use `SqlTypeUtil#isArray` to decide if a type is `ARRAY` type.